File: go/24swapNodesInPairs/24.两两交换链表中的节点.go
/*
 * @lc app=leetcode.cn id=24 lang=golang
 *
 * [24] 两两交换链表中的节点
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func swapPairs(head *ListNode) *ListNode {
	pre := &ListNode{Val:-1}
	dummy := pre
	pre.Next = head
	for ; pre.Next != nil && pre.Next.Next != nil; {
		a := pre.Next
		b := a.Next
		pre.Next, a.Next, b.Next = b, b.Next, a
		pre = a
	}
	return dummy.Next
}
// @lc code=end

File: java/24swapPairs/24.两两交换链表中的节点.java
/*
 * @lc app=leetcode.cn id=24 lang=java
 *
 * [24] 两两交换链表中的节点
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        if(head == null || head.next == null) {
            return head;
        }
        ListNode first = head;
        ListNode second = head.next;
        ListNode pre = null;
        ListNode newHead = second;
        while(second != null) {
            ListNode temp = second.next;
            second.next = first;
            first.next = temp;
            
            if(pre != null) {
                pre.next = second;
            }
            pre = first;

            first = temp;
            if(first != null) {
                second = first.next;
            } else {
                break;
            }
        }
        return newHead;
    }
}
// @lc code=end

